File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/reader/HashObjectTest.java
package com.hedera.mirror.importer.reader;

/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import static com.hedera.mirror.importer.domain.DigestAlgorithm.SHA384;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.*;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.Arrays;
import java.util.Objects;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import com.hedera.mirror.importer.TestUtils;
import com.hedera.mirror.importer.exception.InvalidStreamFileException;

class HashObjectTest {

    @ParameterizedTest
    @CsvSource({
            "1226, 1, , 0, false",
            "1228, 1, , 0, false",
            "1226, 2, , 0, false",
            "1226, 1, 2, 0, true",
            "1226, 1, , 1, true",
            "1226, 1, , 49, true",
            "1226, 1, , 53, true",
            "1226, 1, , 57, true",
            "1226, 1, , 61, true",
    })
    void newHashObject(long classId, int classVersion, Integer digestType, int bytesToTruncate, boolean expectThrown) throws IOException {
        try (ByteArrayOutputStream bos = new ByteArrayOutputStream(); DataOutputStream dos = new DataOutputStream(bos)) {
            // given
            dos.writeLong(classId);
            dos.writeInt(classVersion);
            digestType = Objects.requireNonNullElseGet(digestType, SHA384::getType);
            dos.writeInt(digestType);
            dos.writeInt(SHA384.getSize());
            byte[] hash = TestUtils.generateRandomByteArray(SHA384.getSize());
            dos.write(hash);

            byte[] data = bos.toByteArray();
            if (bytesToTruncate > 0) {
                data = Arrays.copyOfRange(data, 0, data.length - bytesToTruncate);
            }

            try (ValidatedDataInputStream dis = new ValidatedDataInputStream(new ByteArrayInputStream(data), "test")) {
                // when, then
                if (expectThrown) {
                    assertThrows(InvalidStreamFileException.class, () -> new HashObject(dis, SHA384));
                } else {
                    HashObject expected = new HashObject(classId, classVersion, SHA384.getType(), hash);
                    HashObject actual = new HashObject(dis, "testfile", SHA384);
                    assertThat(actual).isEqualTo(expected);
                }
            }
        }
    }
}

File: hedera-mirror-rest/__tests__/stream/hashObject.test.js
/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

'use strict';

const HashObject = require('../../stream/hashObject');

describe('HashObject', () => {
  let buffer;
  const classId = BigInt.asIntN(64, BigInt('0xf422da83a251741e'));
  const classVersion = 1;
  const digestType = 0x58ff811b;
  const hash = new Array(48).fill(0xef);

  beforeEach(() => {
    buffer = Buffer.from(
      [].concat(
        [0xf4, 0x22, 0xda, 0x83, 0xa2, 0x51, 0x74, 0x1e], // classId
        [0, 0, 0, 1], // classVersion
        [0x58, 0xff, 0x81, 0x1b], // digest type, sha-384
        [0, 0, 0, 0x30], // length, for sha-384, 48 bytes
        hash // 48-byte hash
      )
    );
  });

  test('getLength', () => {
    const expected = buffer.length;
    const hashObject = new HashObject(buffer);
    expect(hashObject.getLength()).toEqual(expected);
  });

  test('classId', () => {
    const hashObject = new HashObject(buffer);
    expect(hashObject.classId).toEqual(classId);
  });

  test('classVersion', () => {
    const hashObject = new HashObject(buffer);
    expect(hashObject.classVersion).toEqual(classVersion);
  });

  test('digestType', () => {
    const hashObject = new HashObject(buffer);
    expect(hashObject.digestType).toEqual(digestType);
  });

  test('hash', () => {
    const hashObject = new HashObject(buffer);
    expect(hashObject.hash).toEqual(Buffer.from(hash));
  });

  test('truncated buffer', () => {
    expect(() => new HashObject(buffer.slice(0, buffer.length - 4))).toThrowErrorMatchingSnapshot();
  });
});
